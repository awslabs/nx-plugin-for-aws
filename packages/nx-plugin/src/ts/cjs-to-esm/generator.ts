/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree, getProjects } from '@nx/devkit';
import { CjsToEsmGeneratorSchema } from './schema';
import { tsquery } from '@phenomnomnominal/tsquery';
import * as ts from 'typescript';
import { minimatch } from 'minimatch';
// Replaces "module.exports = value" with "export default value"
const replaceModuleExportsWithExportDefault = (
  sourceFile: ts.SourceFile
): ts.SourceFile => {
  const transformer =
    <T extends ts.Node>(context: ts.TransformationContext) =>
    (rootNode: T) => {
      function visit(node: ts.Node): ts.Node {
        if (ts.isExpressionStatement(node)) {
          const expr = node.expression;
          if (
            ts.isBinaryExpression(expr) &&
            ts.isPropertyAccessExpression(expr.left) &&
            ts.isIdentifier(expr.left.expression) &&
            expr.left.expression.text === 'module' &&
            ts.isIdentifier(expr.left.name) &&
            expr.left.name.text === 'exports' &&
            expr.operatorToken.kind === ts.SyntaxKind.EqualsToken
          ) {
            // Create an export default statement
            return context.factory.createExportAssignment(
              undefined,
              undefined,
              expr.right
            );
          }
        }
        return ts.visitEachChild(node, visit, context);
      }
      return ts.visitNode(rootNode, visit);
    };
  const result = ts.transform(sourceFile, [transformer]);
  return result.transformed[0] as ts.SourceFile;
};
// Replaces "const x = require('y')" with "import x from 'y'"
const replaceRequires = (sourceFile: ts.SourceFile): ts.SourceFile => {
  const transformer =
    <T extends ts.Node>(context: ts.TransformationContext) =>
    (rootNode: T) => {
      function visit(node: ts.Node): ts.Node {
        if (ts.isVariableStatement(node)) {
          const declaration = node.declarationList.declarations[0];
          if (declaration && ts.isVariableDeclaration(declaration)) {
            const initializer = declaration.initializer;
            if (
              initializer &&
              ts.isCallExpression(initializer) &&
              ts.isIdentifier(initializer.expression) &&
              initializer.expression.text === 'require' &&
              initializer.arguments.length === 1
            ) {
              const modulePath = initializer.arguments[0];
              if (ts.isStringLiteral(modulePath)) {
                // Handle destructuring pattern
                if (ts.isObjectBindingPattern(declaration.name)) {
                  const importSpecifiers = declaration.name.elements
                    .map((element) => {
                      if (
                        ts.isBindingElement(element) &&
                        ts.isIdentifier(element.name)
                      ) {
                        return context.factory.createImportSpecifier(
                          false,
                          undefined,
                          element.name
                        );
                      }
                      return undefined;
                    })
                    .filter(
                      (spec): spec is ts.ImportSpecifier => spec !== undefined
                    );
                  return context.factory.createImportDeclaration(
                    undefined,
                    context.factory.createImportClause(
                      false,
                      undefined,
                      context.factory.createNamedImports(importSpecifiers)
                    ),
                    modulePath,
                    undefined
                  );
                }
                // Handle simple require
                if (ts.isIdentifier(declaration.name)) {
                  return context.factory.createImportDeclaration(
                    undefined,
                    context.factory.createImportClause(
                      false,
                      declaration.name,
                      undefined
                    ),
                    modulePath,
                    undefined
                  );
                }
              }
            }
          }
        }
        return ts.visitEachChild(node, visit, context);
      }
      return ts.visitNode(rootNode, visit);
    };
  const result = ts.transform(sourceFile, [transformer]);
  return result.transformed[0] as ts.SourceFile;
};
// Replaces requires not directly assigned to variables eg
// "const x = require('y').foo()" with
// import _import0 from 'y'
// const x = _import0.foo()
const replaceInlineRequires = (sourceFile: ts.SourceFile): ts.SourceFile => {
  let importCounter = 0;
  const imports: ts.ImportDeclaration[] = [];
  const importMap = new Map<string, string>();
  const transformer =
    <T extends ts.Node>(context: ts.TransformationContext) =>
    (rootNode: T) => {
      function visit(node: ts.Node): ts.Node {
        if (
          ts.isCallExpression(node) &&
          ts.isIdentifier(node.expression) &&
          node.expression.text === 'require' &&
          node.arguments.length === 1 &&
          ts.isStringLiteral(node.arguments[0])
        ) {
          const modulePath = node.arguments[0];
          const modulePathText = modulePath.text;
          // Check if we already have an import for this module
          if (!importMap.has(modulePathText)) {
            const importName = `_import${importCounter++}`;
            importMap.set(modulePathText, importName);
            // Create import declaration
            imports.push(
              context.factory.createImportDeclaration(
                undefined,
                context.factory.createImportClause(
                  false,
                  context.factory.createIdentifier(importName),
                  undefined
                ),
                modulePath,
                undefined
              )
            );
          }
          // Replace require() with the import reference
          return context.factory.createIdentifier(
            importMap.get(modulePathText)
          );
        }
        return ts.visitEachChild(node, visit, context);
      }
      // First pass to collect all requires and create imports
      const visited = ts.visitNode(rootNode, visit);
      // If we found any inline requires, prepend the imports to the source file
      if (imports.length > 0 && ts.isSourceFile(visited)) {
        return context.factory.updateSourceFile(visited, [
          ...imports,
          ...visited.statements,
        ]);
      }
      return visited;
    };
  const result = ts.transform(sourceFile, [transformer]);
  return result.transformed[0] as ts.SourceFile;
};
export function cjsToEsm(
  tree: Tree,
  relativePathInTree: string,
  options?: {
    include?: string[];
    exclude?: string[];
  }
) {
  const includePatterns = options?.include ?? ['**/*.{js,ts}'];
  const excludePatterns =
    options?.exclude ??
    ['node_modules', 'dist', 'build', 'tmp', '.nx'].flatMap((p) => [
      p,
      `**/${p}/**/*`,
      `${p}/**/*`,
    ]);
  // Process each TypeScript/JavaScript file in the project
  const processFile = (filePath: string) => {
    const content = tree.read(filePath, 'utf-8');
    if (!content) return;
    // Parse the source file
    const originalSourceFile = tsquery.ast(content);
    let sourceFile = originalSourceFile;
    sourceFile = replaceModuleExportsWithExportDefault(sourceFile);
    sourceFile = replaceRequires(sourceFile);
    sourceFile = replaceInlineRequires(sourceFile);
    const printer = ts.createPrinter({
      removeComments: false,
      newLine: ts.NewLineKind.LineFeed,
    });
    tree.write(
      filePath,
      printer.printNode(ts.EmitHint.Unspecified, sourceFile, originalSourceFile)
    );
  };
  // Process all files in the given directory (recursively)
  const processDirectory = (dirPath: string) => {
    const entries = tree.children(dirPath);
    entries.forEach((entry) => {
      const fullPath = `${dirPath}/${entry}`;
      const relativePath = fullPath.replace(`${relativePathInTree}/`, '');
      // Return early for exclude to ensure we don't traverse node_modules etc!
      const shouldExclude = excludePatterns.some((pattern) =>
        minimatch(relativePath, pattern)
      );
      if (shouldExclude) {
        return;
      }
      if (tree.isFile(fullPath)) {
        const shouldInclude = includePatterns.some((pattern) =>
          minimatch(relativePath, pattern)
        );
        if (shouldInclude) {
          processFile(fullPath);
        }
      } else {
        processDirectory(fullPath);
      }
    });
  };
  // Start processing from the given directories
  processDirectory(relativePathInTree);
}
export async function cjsToEsmGenerator(
  tree: Tree,
  options: CjsToEsmGeneratorSchema
) {
  // Get the project configuration
  const projects = getProjects(tree);
  const project = projects.get(options.project);
  if (!project) {
    throw new Error(`Project ${options.project} not found`);
  }
  cjsToEsm(tree, project.root, {
    include: options.include ? [options.include] : undefined,
    exclude: options.exclude ? [options.exclude] : undefined,
  });
}
export default cjsToEsmGenerator;
