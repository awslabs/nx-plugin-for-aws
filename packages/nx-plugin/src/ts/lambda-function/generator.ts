/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  GeneratorCallback,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  ProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { TsLambdaFunctionGeneratorSchema } from './schema';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from '../../utils/shared-constructs-constants';
import { toClassName, toKebabCase, pascalCase } from '../../utils/names';
import { addStarExport } from '../../utils/ast';
import { formatFilesInSubtree } from '../../utils/format';
import { sortObjectKeys } from '../../utils/object';
import {
  NxGeneratorInfo,
  getGeneratorInfo,
  readProjectConfigurationUnqualified,
} from '../../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../../utils/metrics';
import { withVersions } from '../../utils/versions';
import camelCase from 'lodash.camelcase';
import { TS_HANDLER_RETURN_TYPES } from './io';
import { addEsbuildBundleTarget } from '../../utils/esbuild';

export const TS_LAMBDA_FUNCTION_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

/**
 * Generates a TypeScript Lambda Function to add to a TypeScript project
 */
export const tsLambdaFunctionGenerator = async (
  tree: Tree,
  schema: TsLambdaFunctionGeneratorSchema,
): Promise<GeneratorCallback> => {
  const projectConfig = readProjectConfigurationUnqualified(
    tree,
    schema.project,
  );

  const tsconfigPath = joinPathFragments(projectConfig.root, 'tsconfig.json');

  // Check if the project has a tsconfig.json file
  if (!tree.exists(tsconfigPath)) {
    throw new Error(
      `This generator does not support selected project ${schema.project}. The project must be a typescript project (ie contain a tsconfig.json)`,
    );
  }

  if (!projectConfig.sourceRoot) {
    throw new Error(
      `This project does not have a source root. Please add a source root to the project configuration before running this generator.`,
    );
  }

  const dir = projectConfig.root;
  const projectNameWithOutScope = projectConfig.name.split('/').pop();
  const projectNameKebabCase = toKebabCase(projectNameWithOutScope);
  const functionNameKebabCase = toKebabCase(schema.functionName);

  const constructFunctionNameKebabCase = `${projectNameKebabCase}-${functionNameKebabCase}`;
  const constructFunctionClassName = toClassName(
    constructFunctionNameKebabCase,
  );
  const lambdaFunctionCamelCase = camelCase(schema.functionName);
  const lambdaFunctionClassName = pascalCase(schema.functionName);
  const lambdaFunctionKebabCase = toKebabCase(schema.functionName);

  const functionPath = joinPathFragments(
    projectConfig.sourceRoot,
    schema.functionPath ?? '',
    `${lambdaFunctionKebabCase}.ts`,
  );

  // Check that the project does not already have a lambda handler
  if (tree.exists(functionPath)) {
    throw new Error(
      `This project already has a lambda function with the name ${functionNameKebabCase}. Please remove the lambda function before running this generator or use a different name.`,
    );
  }

  await sharedConstructsGenerator(tree);

  // Add bundle-<name> target for this specific lambda function
  const bundleTargetName = `bundle-${lambdaFunctionKebabCase}`;

  const enhancedOptions = {
    ...schema,
    dir,
    constructFunctionClassName,
    constructFunctionNameKebabCase,
    lambdaFunctionCamelCase,
    lambdaFunctionClassName,
    lambdaFunctionKebabCase,
    bundleTargetName,
    returnType: TS_HANDLER_RETURN_TYPES[schema.eventSource],
  };

  // Add a bundle target for the function
  addEsbuildBundleTarget(projectConfig, {
    bundleTargetName,
    targetFilePath: functionPath,
    extraEsbuildArgs: '--external:@aws-sdk/*',
  });

  projectConfig.targets = sortObjectKeys(projectConfig.targets);
  updateProjectConfiguration(tree, projectConfig.name, projectConfig);

  // Generate the lambda handler file
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'handler'),
    joinPathFragments(projectConfig.sourceRoot, schema.functionPath ?? ''),
    enhancedOptions,
    { overwriteStrategy: OverwriteStrategy.KeepExisting },
  );

  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', SHARED_CONSTRUCTS_DIR, 'src', 'app'),
    joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src', 'app'),
    enhancedOptions,
    { overwriteStrategy: OverwriteStrategy.KeepExisting },
  );

  addStarExport(
    tree,
    joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'index.ts',
    ),
    './lambda-functions/index.js',
  );
  addStarExport(
    tree,
    joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'lambda-functions',
      'index.ts',
    ),
    `./${constructFunctionNameKebabCase}.js`,
  );

  // Ensure common constructs builds after our lambda function project
  updateJson(
    tree,
    joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
    (config: ProjectConfiguration) => {
      if (!config.targets) {
        config.targets = {};
      }
      if (!config.targets.build) {
        config.targets.build = {};
      }
      config.targets.build.dependsOn = [
        ...(config.targets.build.dependsOn ?? []).filter(
          (t) => t !== `${projectConfig.name}:build`,
        ),
        `${projectConfig.name}:build`,
      ];
      return config;
    },
  );

  addDependenciesToPackageJson(
    tree,
    withVersions([
      '@aws-lambda-powertools/tracer',
      '@aws-lambda-powertools/logger',
      '@aws-lambda-powertools/metrics',
      '@aws-lambda-powertools/parser',
      '@middy/core',
      'zod',
    ]),
    withVersions(['esbuild', '@types/aws-lambda']),
  );

  await addGeneratorMetricsIfApplicable(tree, [
    TS_LAMBDA_FUNCTION_GENERATOR_INFO,
  ]);

  await formatFilesInSubtree(tree);

  return () => {
    installPackagesTask(tree);
  };
};

export default tsLambdaFunctionGenerator;
