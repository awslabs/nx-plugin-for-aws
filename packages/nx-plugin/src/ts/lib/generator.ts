/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  GeneratorCallback,
  NxJsonConfiguration,
  OverwriteStrategy,
  Tree,
  generateFiles,
  getPackageManagerCommand,
  installPackagesTask,
  joinPathFragments,
  readProjectConfiguration,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { TsLibGeneratorSchema } from './schema';
import { libraryGenerator } from '@nx/js';
import { getNpmScopePrefix } from '../../utils/npm-scope';
import { configureTsProject } from './ts-project-utils';
import { toKebabCase } from '../../utils/names';
import { relative } from 'path';
import { formatFilesInSubtree } from '../../utils/format';
import { sortObjectKeys } from '../../utils/nx';
export interface TsLibDetails {
  /**
   * Full package name including scope (eg @foo/bar)
   */
  readonly fullyQualifiedName: string;
  /**
   * Directory of the library relative to the root
   */
  readonly dir: string;
}
/**
 * Returns details about the TS library to be created
 */
export const getTsLibDetails = (
  tree: Tree,
  schema: TsLibGeneratorSchema,
): TsLibDetails => {
  const scope = getNpmScopePrefix(tree);
  const normalizedName = toKebabCase(schema.name);
  const fullyQualifiedName = `${scope}${normalizedName}`;
  const dir = joinPathFragments(
    toKebabCase(schema.directory) ?? '.',
    toKebabCase(schema.subDirectory) ?? normalizedName,
  );
  return { dir, fullyQualifiedName };
};
/**
 * Generates a typescript library
 */
export const tsLibGenerator = async (
  tree: Tree,
  schema: TsLibGeneratorSchema,
): Promise<GeneratorCallback> => {
  const { fullyQualifiedName, dir } = getTsLibDetails(tree, schema);
  await libraryGenerator(tree, {
    ...schema,
    name: toKebabCase(schema.name),
    directory: dir,
    skipPackageJson: true,
    bundler: 'tsc', // TODO: consider supporting others
    linter: 'eslint',
    unitTestRunner: 'vitest',
  });
  // Replace with simpler sample source code
  tree.delete(joinPathFragments(dir, 'src'));
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    joinPathFragments(dir),
    {
      fullyQualifiedName,
      pkgMgrCmd: getPackageManagerCommand().exec,
    },
    {
      overwriteStrategy: OverwriteStrategy.KeepExisting,
    },
  );
  configureTsProject(tree, {
    dir,
    fullyQualifiedName,
  });

  const projectConfiguration = readProjectConfiguration(
    tree,
    fullyQualifiedName,
  );
  const targets = projectConfiguration.targets;

  targets['compile'] = {
    executor: 'nx:run-commands',
    outputs: [`{workspaceRoot}/dist/${dir}/tsc`],
    options: {
      command: 'tsc --build tsconfig.lib.json',
      cwd: '{projectRoot}',
    },
  };
  targets['build'] = {
    dependsOn: ['lint', 'compile', 'test'],
  };
  targets['test'] = {
    executor: '@nx/vite:test',
    outputs: ['{options.reportsDirectory}'],
    options: {
      reportsDirectory: joinPathFragments(
        relative(joinPathFragments(tree.root, dir), tree.root),
        'coverage',
        dir,
      ),
    },
  };
  delete targets['lint'];
  projectConfiguration.targets = sortObjectKeys(targets);

  updateProjectConfiguration(tree, fullyQualifiedName, projectConfiguration);

  updateJson(tree, 'nx.json', (nxJson: NxJsonConfiguration) => {
    nxJson.plugins = nxJson.plugins.map((p) => {
      if (
        typeof p !== 'string' &&
        p.plugin === '@nx/js/typescript' &&
        p.options?.['build']
      ) {
        p.options['build'].targetName = 'compile';
      }
      return p;
    });
    return nxJson;
  });

  await formatFilesInSubtree(tree);

  return () => {
    if (!schema.skipInstall) {
      installPackagesTask(tree);
    }
  };
};
export default tsLibGenerator;
