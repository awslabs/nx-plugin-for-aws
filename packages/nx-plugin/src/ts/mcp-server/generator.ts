/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  GeneratorCallback,
  OverwriteStrategy,
  Tree,
  addDependenciesToPackageJson,
  generateFiles,
  installPackagesTask,
  joinPathFragments,
  readProjectConfiguration,
  updateProjectConfiguration,
} from '@nx/devkit';
import { TsMcpServerGeneratorSchema } from './schema';
import { NxGeneratorInfo, getGeneratorInfo } from '../../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../../utils/metrics';
import { formatFilesInSubtree } from '../../utils/format';
import tsProjectGenerator, { getTsLibDetails } from '../lib/generator';
import { withVersions } from '../../utils/versions';

export const TS_MCP_SERVER_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

export const tsMcpServerGenerator = async (
  tree: Tree,
  options: TsMcpServerGeneratorSchema,
): Promise<GeneratorCallback> => {
  // Generate a TypeScript library
  await tsProjectGenerator(tree, options);

  const { fullyQualifiedName } = getTsLibDetails(tree, options);
  const project = readProjectConfiguration(tree, fullyQualifiedName);

  // Add dependencies
  const deps = withVersions(['@modelcontextprotocol/sdk', 'zod']);
  const devDeps = withVersions(['tsx', 'esbuild']);
  addDependenciesToPackageJson(tree, deps, devDeps);

  // Generate example server
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    project.root,
    {
      name: options.name,
      dir: project.root,
    },
    { overwriteStrategy: OverwriteStrategy.Overwrite },
  );

  // Add build targets
  updateProjectConfiguration(tree, fullyQualifiedName, {
    ...project,
    targets: {
      ...project?.targets,
      // Bundle with esbuild
      bundle: {
        executor: 'nx:run-commands',
        options: {
          commands: [
            `esbuild "${joinPathFragments(project.sourceRoot, 'index.ts')}" --bundle --platform=node --format=esm --loader:.md=text --outfile=${joinPathFragments('dist', project.root, 'bundle', 'index.js')}`,
          ],
        },
      },
      // Dev task for bundling whenever changes are made
      dev: {
        executor: 'nx:run-commands',
        options: {
          commands: [
            `nx watch --projects=${fullyQualifiedName} --includeDependentProjects -- nx run ${fullyQualifiedName}:bundle`,
          ],
        },
      },
      // Ensure bundle runs as part of build
      build: {
        ...project?.targets?.build,
        dependsOn: [...(project?.targets?.build?.dependsOn ?? []), 'bundle'],
      },
    },
  });

  await addGeneratorMetricsIfApplicable(tree, [TS_MCP_SERVER_GENERATOR_INFO]);

  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
};

export default tsMcpServerGenerator;
