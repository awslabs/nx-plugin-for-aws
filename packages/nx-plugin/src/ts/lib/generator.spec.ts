/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from 'nx/src/devkit-testing-exports';
import { tsLibGenerator, getTsLibDetails } from './generator';
describe('ts lib generator', () => {
  let tree: Tree;
  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });
  it('should generate library with default options', async () => {
    await tsLibGenerator(tree, {
      name: 'test-lib',
      unitTestRunner: 'vitest',
      skipInstall: true,
    });
    // Verify directory structure
    expect(tree.exists('test-lib')).toBeTruthy();
    expect(tree.exists('test-lib/src')).toBeTruthy();
    expect(tree.exists('test-lib/src/index.ts')).toBeTruthy();
    expect(tree.exists('test-lib/tsconfig.json')).toBeTruthy();
    expect(tree.exists('test-lib/project.json')).toBeTruthy();
    // Create snapshots of generated files
    expect(tree.read('test-lib/src/index.ts', 'utf-8')).toMatchSnapshot(
      'index.ts'
    );
    expect(tree.read('test-lib/tsconfig.json', 'utf-8')).toMatchSnapshot(
      'tsconfig.json'
    );
    expect(tree.read('test-lib/project.json', 'utf-8')).toMatchSnapshot(
      'project.json'
    );
  });
  it('should generate library with custom directory', async () => {
    await tsLibGenerator(tree, {
      name: 'test-lib',
      directory: 'libs',
      unitTestRunner: 'vitest',
      skipInstall: true,
    });
    // Verify directory structure
    expect(tree.exists('libs/test-lib')).toBeTruthy();
    expect(tree.exists('libs/test-lib/src')).toBeTruthy();
    expect(tree.exists('libs/test-lib/src/index.ts')).toBeTruthy();
    // Create snapshots of generated files
    expect(tree.read('libs/test-lib/src/index.ts', 'utf-8')).toMatchSnapshot(
      'custom-dir-index.ts'
    );
    expect(tree.read('libs/test-lib/tsconfig.json', 'utf-8')).toMatchSnapshot(
      'custom-dir-tsconfig.json'
    );
    expect(tree.read('libs/test-lib/project.json', 'utf-8')).toMatchSnapshot(
      'custom-dir-project.json'
    );
  });
  it('should generate library with custom scope', async () => {
    await tsLibGenerator(tree, {
      name: 'test-lib',
      scope: '@custom',
      unitTestRunner: 'vitest',
      skipInstall: true,
    });
    const details = getTsLibDetails(tree, {
      name: 'test-lib',
      scope: '@custom',
    });
    expect(details.fullyQualifiedName).toBe('@custom/test-lib');
    expect(tree.exists('test-lib/project.json')).toBeTruthy();
    expect(tree.read('test-lib/project.json', 'utf-8')).toMatchSnapshot(
      'scoped-project.json'
    );
  });
  it('should generate library with subdirectory', async () => {
    await tsLibGenerator(tree, {
      name: 'test-lib',
      subDirectory: 'test-lib',
      directory: 'feature',
      unitTestRunner: 'vitest',
      skipInstall: true,
    });
    // Verify directory structure
    expect(tree.exists('feature/test-lib')).toBeTruthy();
    expect(tree.exists('feature/test-lib/src')).toBeTruthy();
    expect(tree.exists('feature/test-lib/src/index.ts')).toBeTruthy();
    // Create snapshots of generated files
    expect(tree.read('feature/test-lib/src/index.ts', 'utf-8')).toMatchSnapshot(
      'subdir-index.ts'
    );
    expect(
      tree.read('feature/test-lib/tsconfig.json', 'utf-8')
    ).toMatchSnapshot('subdir-tsconfig.json');
    expect(tree.read('feature/test-lib/project.json', 'utf-8')).toMatchSnapshot(
      'subdir-project.json'
    );
  });
});
