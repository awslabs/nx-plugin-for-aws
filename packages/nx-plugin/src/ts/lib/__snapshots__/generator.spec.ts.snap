// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ts lib generator > should configure named inputs in nx.json 1`] = `
"{
  "affected": {
    "defaultBase": "main"
  },
  "targetDefaults": {
    "build": {
      "cache": true,
      "inputs": ["default"]
    },
    "lint": {
      "cache": true,
      "configurations": {
        "fix": {
          "fix": true
        }
      },
      "inputs": [
        "default",
        "{workspaceRoot}/eslint.config.mjs",
        "{projectRoot}/eslint.config.mjs"
      ]
    },
    "@nx/eslint:lint": {
      "cache": true,
      "inputs": [
        "default",
        "{workspaceRoot}/.eslintrc.json",
        "{workspaceRoot}/.eslintignore",
        "{workspaceRoot}/eslint.config.mjs"
      ]
    },
    "@nx/vite:test": {
      "cache": true,
      "inputs": ["default", "^default"],
      "configurations": {
        "update-snapshot": {
          "args": "--update"
        }
      }
    },
    "test": {
      "dependsOn": ["^build"],
      "inputs": ["default"]
    },
    "compile": {
      "cache": true,
      "inputs": ["default"]
    }
  },
  "plugins": [
    {
      "plugin": "@nx/js/typescript",
      "options": {
        "typecheck": {
          "targetName": "typecheck"
        },
        "build": {
          "targetName": "compile",
          "configName": "tsconfig.lib.json",
          "buildDepsName": "build-deps",
          "watchDepsName": "watch-deps"
        }
      }
    },
    {
      "plugin": "@nx/eslint/plugin",
      "options": {
        "targetName": "lint"
      }
    }
  ],
  "namedInputs": {
    "default": [
      {
        "dependentTasksOutputFiles": "**/*",
        "transitive": true
      }
    ]
  }
}
"
`;

exports[`ts lib generator > should generate library with custom directory > custom-dir-index.ts 1`] = `
"export function hello() {
  return 'Hello!';
}
"
`;

exports[`ts lib generator > should generate library with custom directory > custom-dir-project.json 1`] = `
"{
  "name": "@proj/test-lib",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/test-lib/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "build": {
      "dependsOn": ["lint", "compile", "test"]
    },
    "compile": {
      "executor": "nx:run-commands",
      "outputs": ["{workspaceRoot}/dist/libs/test-lib/tsc"],
      "options": {
        "command": "tsc --build tsconfig.lib.json",
        "cwd": "{projectRoot}"
      }
    },
    "test": {
      "executor": "@nx/vite:test",
      "outputs": ["{options.reportsDirectory}"],
      "options": {
        "reportsDirectory": "../../coverage/libs/test-lib"
      }
    }
  }
}
"
`;

exports[`ts lib generator > should generate library with custom directory > custom-dir-tsconfig.json 1`] = `
"{
  "extends": "../../tsconfig.base.json",
  "files": [],
  "include": [],
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ],
  "compilerOptions": {}
}
"
`;

exports[`ts lib generator > should generate library with default options > eslint.config.mjs 1`] = `
"import baseConfig from '../eslint.config.mjs';

export default [
  ...baseConfig,
  {
    files: ['**/*.json'],
    rules: {
      '@nx/dependency-checks': [
        'warn',
        {
          ignoredFiles: [
            '{projectRoot}/eslint.config.{js,cjs,mjs}',
            '{projectRoot}/vite.config.{js,ts,mjs,mts}',
          ],
        },
      ],
    },
    languageOptions: {
      parser: await import('jsonc-eslint-parser'),
    },
  },
];
"
`;

exports[`ts lib generator > should generate library with default options > index.ts 1`] = `
"export function hello() {
  return 'Hello!';
}
"
`;

exports[`ts lib generator > should generate library with default options > project.json 1`] = `
"{
  "name": "@proj/test-lib",
  "$schema": "../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "test-lib/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "build": {
      "dependsOn": ["lint", "compile", "test"]
    },
    "compile": {
      "executor": "nx:run-commands",
      "outputs": ["{workspaceRoot}/dist/test-lib/tsc"],
      "options": {
        "command": "tsc --build tsconfig.lib.json",
        "cwd": "{projectRoot}"
      }
    },
    "test": {
      "executor": "@nx/vite:test",
      "outputs": ["{options.reportsDirectory}"],
      "options": {
        "reportsDirectory": "../coverage/test-lib"
      }
    }
  }
}
"
`;

exports[`ts lib generator > should generate library with default options > tsconfig.json 1`] = `
"{
  "extends": "../tsconfig.base.json",
  "files": [],
  "include": [],
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ],
  "compilerOptions": {}
}
"
`;

exports[`ts lib generator > should generate library with subdirectory > subdir-index.ts 1`] = `
"export function hello() {
  return 'Hello!';
}
"
`;

exports[`ts lib generator > should generate library with subdirectory > subdir-project.json 1`] = `
"{
  "name": "@proj/test-lib",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "feature/test-lib/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "build": {
      "dependsOn": ["lint", "compile", "test"]
    },
    "compile": {
      "executor": "nx:run-commands",
      "outputs": ["{workspaceRoot}/dist/feature/test-lib/tsc"],
      "options": {
        "command": "tsc --build tsconfig.lib.json",
        "cwd": "{projectRoot}"
      }
    },
    "test": {
      "executor": "@nx/vite:test",
      "outputs": ["{options.reportsDirectory}"],
      "options": {
        "reportsDirectory": "../../coverage/feature/test-lib"
      }
    }
  }
}
"
`;

exports[`ts lib generator > should generate library with subdirectory > subdir-tsconfig.json 1`] = `
"{
  "extends": "../../tsconfig.base.json",
  "files": [],
  "include": [],
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ],
  "compilerOptions": {}
}
"
`;
